File: Programmers/Level_1/소수만들기.cpp
#include <bits/stdc++.h>
using namespace std;
#define MAX 3001

bool isPrime[MAX];
int makeSum[3];
int answer;

void sosu() {
    for (int i = 2; i < MAX; i++) {
        isPrime[i] = true;
    }
    for (int i = 2; i < sqrt(MAX); i++) {
        if (!isPrime[i]) {
            continue;
        }
        for (int j = i+i; j < MAX; j += i) {
            isPrime[j] = false;
        }
    }
}

void DFS(vector<int> &tmpNums, int idx, int cnt) {
    // 종료조건 
    if (cnt == 3) {
        if (isPrime[makeSum[0] + makeSum[1] + makeSum[2]]) {
            answer++;
        }
        return;
    }
    for (int i = idx; i < tmpNums.size(); i++) {
        makeSum[cnt] = tmpNums[i];
        DFS(tmpNums, i+1, cnt+1);
    }
}

int solution(vector<int> &nums) {
    sosu();
    DFS(nums, 0, 0);
    return answer;
}

int main(int argc, char const *argv[]) {
    vector<int> nums = {1,2,3,4};
    cout << solution(nums) << endl;
    return 0;
}

File: Programmers/Level_1/Swift_Codes/Swift_Codes/소수만들기.swift
//
//  소수만들기.swift
//  Swift_Codes
//
//  Created by InJe Choi on 2021/09/13.
//

import Foundation

let MAX = 3001
var isPrime = [Bool](repeating: true, count: MAX)
var makeSum = [Int](repeating: 0, count: 3)
var answer = 0

func sosu() {
    isPrime[0...1] = [false, false]
    for i in stride(from: 2, to: Int(sqrt(Double(MAX))), by: 1) {
        if isPrime[i] == false {
            continue
        }
        for j in stride(from: i+i, to: MAX, by: i) {
            isPrime[j] = false
        }
    }
}

func DFS(_ tmpNums:[Int], _ idx:Int, _ cnt:Int) {
    if cnt == 3 {
        var sum = 0
        sum += makeSum.reduce(0, +)
        if isPrime[sum] == true {
            answer += 1
        }
        return
    }
    for i in stride(from: idx, to: tmpNums.count, by: 1) {
        makeSum[cnt] = tmpNums[i]
        DFS(tmpNums, i+1, cnt+1)
    }
}

func solution(_ nums:[Int]) -> Int {
    sosu()
    DFS(nums, 0, 0)
    return answer
}
